File: src/DesignPattern/Behavioral/TemplateMethod/TM-go/templateMethod.go
package main

import "fmt"

// AbstractClass ...
type AbstractClass interface {
	TemplateMethod()
}

// ConcreteClass ..
type ConcreteClass struct {
	step2 func()
}

// TemplateMethod ...
func (p *ConcreteClass) TemplateMethod() {
	p.step1()
	p.step2()
	p.step3()
}

func (p *ConcreteClass) step1() {
	fmt.Println("step 1 ...")
}

func (p *ConcreteClass) step3() {
	fmt.Println("step 3 ...")
}

// NewTemplateMethod ...
func NewTemplateMethod(f func()) AbstractClass {
	return &ConcreteClass{
		step2: f,
	}
}

File: src/DesignPattern/Behavioral/TemplateMethod/TM-ts/templateMethod.ts
export abstract class AbstractClass {
    public TemplateMethod() {
        this.step1();
        this.step2();
        this.step3();
    }
    protected step1() {
        console.log("step 1");
    }
    protected abstract step2();

    protected step3() {
        console.log("step 3");
    }

}

